aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs

```diff
@@ -0,0 +1,216 @@
+use crate::mock_grpc::MockGrpcServer;
+use anyhow::Context;
+use aptos_indexer_processor_sdk::{
+    aptos_indexer_transaction_stream::TransactionStreamConfig,
+    traits::processor_trait::ProcessorTrait,
+};
+use aptos_protos::{indexer::v1::TransactionsResponse, transaction::v1::Transaction};
+use serde_json::{to_string_pretty, Error as SerdeError};
+use std::{
+    collections::HashMap,
+    fs,
+    path::{Path, PathBuf},
+    time::Duration,
+};
+use tokio::{
+    sync::Mutex,
+    time::{self, Duration as TokioDuration},
+};
+use tokio_retry::{
+    strategy::{jitter, ExponentialBackoff},
+    Retry,
+};
+use url::Url;
+
+const INDEXER_GRPC_DATA_SERVICE_URL: &str = "http://localhost:51254";
+
+pub struct SdkTestContext {
+    pub transaction_batches: Vec<Transaction>,
+}
+
+impl SdkTestContext {
+    pub async fn new(txn_bytes: &[&[u8]]) -> anyhow::Result<Self> {
+        let transaction_batches = txn_bytes
+            .iter()
+            .enumerate()
+            .map(|(idx, txn)| {
+                serde_json::from_slice::<Transaction>(txn).map_err(|err| {
+                    anyhow::anyhow!(
+                        "Failed to parse transaction at index {}: {}",
+                        idx,
+                        format_serde_error(err)
+                    )
+                })
+            })
+            .collect::<Result<Vec<Transaction>, _>>()?;
+
+        let context = SdkTestContext {
+            transaction_batches,
+        };
+
+        // Create mock GRPC transactions and setup the server
+        let transactions = context.transaction_batches.clone();
+        let transactions_response = vec![TransactionsResponse {
+            transactions,
+            ..TransactionsResponse::default()
+        }];
+
+        // Call setup_mock_grpc to start the server and get the port
+        context.setup_mock_grpc(transactions_response, 1).await;
+
+        Ok(context)
+    }
+
+    /// Run the processor and pass user-defined validation logic
+    pub async fn run<F>(
+        &mut self,
+        processor: &impl ProcessorTrait,
+        db_url: &str,
+        txn_version: u64,
+        generate_files: bool, // flag to control file generation
+        output_path: String,  // output path
+        verification_f: F,    // Modified to return a HashMap for multi-table data
+    ) -> anyhow::Result<HashMap<String, serde_json::Value>>
+    // Return HashMap for multi-table results
+    where
+        F: FnOnce(&str) -> anyhow::Result<HashMap<String, serde_json::Value>>
+            + Send
+            + Sync
+            + 'static, // Modified for multi-table verification
+    {
+        let retry_strategy = ExponentialBackoff::from_millis(100).map(jitter).take(5); // Retry up to 5 times
+
+        let timeout_duration = TokioDuration::from_secs(10); // e.g., 5 seconds timeout for each retry
```

Comment:
How often does this fail? 😥

Reply:
after adding dynamic porting, it hasn't failed yet. but will keep it here just in case

Reply:
https://github.com/aptos-labs/aptos-indexer-processor-sdk/pull/70 this should handle.